<!DOCTYPE html><html lang="en" data-theme="light"><head><meta charset="UTF-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0,viewport-fit=cover"><title>240102 Report Draft of DESeq2 WorkFlow | Mengqi's Blog</title><meta name="author" content="Mengqi"><meta name="copyright" content="Mengqi"><meta name="format-detection" content="telephone=no"><meta name="theme-color" content="#ffffff"><meta name="description" content="240102汇报口水稿，仅留档用，无演示，不建议阅读。 相关内容PDF@6010_disk1_gmq: Report&#x2F;240102_DESeq2_WorkFlow.pdfCode@6010_disk1_gmq: Code&#x2F;240102_DESeq2_WorkFlow.RLink: https:&#x2F;&#x2F;master.bioconductor.org&#x2F;packages&#x2F;release&#x2F;">
<meta property="og:type" content="article">
<meta property="og:title" content="240102 Report Draft of DESeq2 WorkFlow">
<meta property="og:url" content="http://www.monig.cn/2024/01/02/240102%20Report%20Draft%20of%20DESeq2%20WorkFlow/index.html">
<meta property="og:site_name" content="Mengqi&#39;s Blog">
<meta property="og:description" content="240102汇报口水稿，仅留档用，无演示，不建议阅读。 相关内容PDF@6010_disk1_gmq: Report&#x2F;240102_DESeq2_WorkFlow.pdfCode@6010_disk1_gmq: Code&#x2F;240102_DESeq2_WorkFlow.RLink: https:&#x2F;&#x2F;master.bioconductor.org&#x2F;packages&#x2F;release&#x2F;">
<meta property="og:locale" content="en_US">
<meta property="og:image" content="https://s2.loli.net/2024/01/14/BlD6LX3CoIqpn4x.jpg">
<meta property="article:published_time" content="2024-01-02T12:30:00.000Z">
<meta property="article:modified_time" content="2024-01-13T12:40:00.000Z">
<meta property="article:author" content="Mengqi">
<meta property="article:tag" content="Not Recommended Reading">
<meta name="twitter:card" content="summary">
<meta name="twitter:image" content="https://s2.loli.net/2024/01/14/BlD6LX3CoIqpn4x.jpg"><link rel="shortcut icon" href="https://s2.loli.net/2024/01/14/ZFXlLMRh3WTbQkE.png"><link rel="canonical" href="http://www.monig.cn/2024/01/02/240102%20Report%20Draft%20of%20DESeq2%20WorkFlow/index.html"><link rel="preconnect" href="//cdn.jsdelivr.net"/><link rel="preconnect" href="//busuanzi.ibruce.info"/><link rel="stylesheet" href="/css/index.css?v=4.12.0"><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@fortawesome/fontawesome-free@6.5.1/css/all.min.css"><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@fancyapps/ui@5.0.32/dist/fancybox/fancybox.min.css" media="print" onload="this.media='all'"><script>const GLOBAL_CONFIG = {
  root: '/',
  algolia: undefined,
  localSearch: undefined,
  translate: undefined,
  noticeOutdate: {"limitDay":365,"position":"top","messagePrev":"It has been","messageNext":"days since the last update, the content of the article may be outdated."},
  highlight: {"plugin":"highlight.js","highlightCopy":true,"highlightLang":true,"highlightHeightLimit":230},
  copy: {
    success: 'Copy Successful',
    error: 'Copy Error',
    noSupport: 'Browser Not Supported'
  },
  relativeDate: {
    homepage: false,
    post: false
  },
  runtime: '',
  dateSuffix: {
    just: 'Just now',
    min: 'minutes ago',
    hour: 'hours ago',
    day: 'days ago',
    month: 'months ago'
  },
  copyright: {"limitCount":150,"languages":{"author":"Author: Mengqi","link":"Link: ","source":"Source: Mengqi's Blog","info":"Copyright is owned by the author. For commercial reprints, please contact the author for authorization. For non-commercial reprints, please indicate the source."}},
  lightbox: 'fancybox',
  Snackbar: undefined,
  infinitegrid: {
    js: 'https://cdn.jsdelivr.net/npm/@egjs/infinitegrid@4.11.0/dist/infinitegrid.min.js',
    buttonText: 'Load More'
  },
  isPhotoFigcaption: false,
  islazyload: false,
  isAnchor: false,
  percent: {
    toc: true,
    rightside: false,
  },
  autoDarkmode: true
}</script><script id="config-diff">var GLOBAL_CONFIG_SITE = {
  title: '240102 Report Draft of DESeq2 WorkFlow',
  isPost: true,
  isHome: false,
  isHighlightShrink: false,
  isToc: true,
  postUpdate: '2024-01-13 20:40:00'
}</script><script>(win=>{
      win.saveToLocal = {
        set: (key, value, ttl) => {
          if (ttl === 0) return
          const now = Date.now()
          const expiry = now + ttl * 86400000
          const item = {
            value,
            expiry
          }
          localStorage.setItem(key, JSON.stringify(item))
        },
      
        get: key => {
          const itemStr = localStorage.getItem(key)
      
          if (!itemStr) {
            return undefined
          }
          const item = JSON.parse(itemStr)
          const now = Date.now()
      
          if (now > item.expiry) {
            localStorage.removeItem(key)
            return undefined
          }
          return item.value
        }
      }
    
      win.getScript = (url, attr = {}) => new Promise((resolve, reject) => {
        const script = document.createElement('script')
        script.src = url
        script.async = true
        script.onerror = reject
        script.onload = script.onreadystatechange = function() {
          const loadState = this.readyState
          if (loadState && loadState !== 'loaded' && loadState !== 'complete') return
          script.onload = script.onreadystatechange = null
          resolve()
        }

        Object.keys(attr).forEach(key => {
          script.setAttribute(key, attr[key])
        })

        document.head.appendChild(script)
      })
    
      win.getCSS = (url, id = false) => new Promise((resolve, reject) => {
        const link = document.createElement('link')
        link.rel = 'stylesheet'
        link.href = url
        if (id) link.id = id
        link.onerror = reject
        link.onload = link.onreadystatechange = function() {
          const loadState = this.readyState
          if (loadState && loadState !== 'loaded' && loadState !== 'complete') return
          link.onload = link.onreadystatechange = null
          resolve()
        }
        document.head.appendChild(link)
      })
    
      win.activateDarkMode = () => {
        document.documentElement.setAttribute('data-theme', 'dark')
        if (document.querySelector('meta[name="theme-color"]') !== null) {
          document.querySelector('meta[name="theme-color"]').setAttribute('content', '#0d0d0d')
        }
      }
      win.activateLightMode = () => {
        document.documentElement.setAttribute('data-theme', 'light')
        if (document.querySelector('meta[name="theme-color"]') !== null) {
          document.querySelector('meta[name="theme-color"]').setAttribute('content', '#ffffff')
        }
      }
      const t = saveToLocal.get('theme')
    
          const isDarkMode = window.matchMedia('(prefers-color-scheme: dark)').matches
          const isLightMode = window.matchMedia('(prefers-color-scheme: light)').matches
          const isNotSpecified = window.matchMedia('(prefers-color-scheme: no-preference)').matches
          const hasNoSupport = !isDarkMode && !isLightMode && !isNotSpecified

          if (t === undefined) {
            if (isLightMode) activateLightMode()
            else if (isDarkMode) activateDarkMode()
            else if (isNotSpecified || hasNoSupport) {
              const now = new Date()
              const hour = now.getHours()
              const isNight = hour <= 6 || hour >= 18
              isNight ? activateDarkMode() : activateLightMode()
            }
            window.matchMedia('(prefers-color-scheme: dark)').addListener(e => {
              if (saveToLocal.get('theme') === undefined) {
                e.matches ? activateDarkMode() : activateLightMode()
              }
            })
          } else if (t === 'light') activateLightMode()
          else activateDarkMode()
        
      const asideStatus = saveToLocal.get('aside-status')
      if (asideStatus !== undefined) {
        if (asideStatus === 'hide') {
          document.documentElement.classList.add('hide-aside')
        } else {
          document.documentElement.classList.remove('hide-aside')
        }
      }
    
      const detectApple = () => {
        if(/iPad|iPhone|iPod|Macintosh/.test(navigator.userAgent)){
          document.documentElement.classList.add('apple')
        }
      }
      detectApple()
    })(window)</script><link rel="stylesheet" href="/css/custom.css" media="defer" onload="this.media='all'"><!-- hexo injector head_end start --><link rel="stylesheet" href="https://cdn.jsdelivr.net/gh/Zfour/Butterfly-double-row-display@1.00/cardlistpost.min.css"/>
<style>#recent-posts > .recent-post-item >.recent-post-info > .article-meta-wrap > .tags:before {content:"\A";
  white-space: pre;}#recent-posts > .recent-post-item >.recent-post-info > .article-meta-wrap > .tags > .article-meta__separator{display:none}</style>
<!-- hexo injector head_end end --><meta name="generator" content="Hexo 6.3.0"></head><body><script>window.paceOptions = {
  restartOnPushState: false
}

document.addEventListener('pjax:send', () => {
  Pace.restart()
})
</script><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/pace-js@1.2.4/themes/blue/pace-theme-minimal.min.css"/><script src="https://cdn.jsdelivr.net/npm/pace-js@1.2.4/pace.min.js"></script><div id="sidebar"><div id="menu-mask"></div><div id="sidebar-menus"><div class="avatar-img is-center"><img src="https://s2.loli.net/2024/01/14/Y51RcC3rJjgFDkS.jpg" onerror="onerror=null;src='/img/friend_404.gif'" alt="avatar"/></div><div class="sidebar-site-data site-data is-center"><a href="/archives/"><div class="headline">Articles</div><div class="length-num">2</div></a><a href="/tags/"><div class="headline">Tags</div><div class="length-num">1</div></a><a href="/categories/"><div class="headline">Categories</div><div class="length-num">1</div></a></div><hr class="custom-hr"/><div class="menus_items"><div class="menus_item"><a class="site-page" href="/"><i class="fa-fw fas fa-home"></i><span> Home</span></a></div><div class="menus_item"><a class="site-page" href="/archives/"><i class="fa-fw fas fa-archive"></i><span> Archives</span></a></div><div class="menus_item"><a class="site-page" href="/tags/"><i class="fa-fw fas fa-tags"></i><span> Tags</span></a></div><div class="menus_item"><a class="site-page" href="/categories/"><i class="fa-fw fas fa-folder-open"></i><span> Categories</span></a></div><div class="menus_item"><a class="site-page" href="/link/"><i class="fa-fw fas fa-link"></i><span> Link</span></a></div></div></div></div><div class="post" id="body-wrap"><header class="not-top-img" id="page-header"><nav id="nav"><span id="blog-info"><a href="/" title="Mengqi's Blog"><span class="site-name">Mengqi's Blog</span></a></span><div id="menus"><div class="menus_items"><div class="menus_item"><a class="site-page" href="/"><i class="fa-fw fas fa-home"></i><span> Home</span></a></div><div class="menus_item"><a class="site-page" href="/archives/"><i class="fa-fw fas fa-archive"></i><span> Archives</span></a></div><div class="menus_item"><a class="site-page" href="/tags/"><i class="fa-fw fas fa-tags"></i><span> Tags</span></a></div><div class="menus_item"><a class="site-page" href="/categories/"><i class="fa-fw fas fa-folder-open"></i><span> Categories</span></a></div><div class="menus_item"><a class="site-page" href="/link/"><i class="fa-fw fas fa-link"></i><span> Link</span></a></div></div><div id="toggle-menu"><a class="site-page" href="javascript:void(0);"><i class="fas fa-bars fa-fw"></i></a></div></div></nav></header><main class="layout" id="content-inner"><div id="post"><div id="post-info"><h1 class="post-title">240102 Report Draft of DESeq2 WorkFlow</h1><div id="post-meta"><div class="meta-firstline"><span class="post-meta-date"><i class="far fa-calendar-alt fa-fw post-meta-icon"></i><span class="post-meta-label">Created</span><time class="post-meta-date-created" datetime="2024-01-02T12:30:00.000Z" title="Created 2024-01-02 20:30:00">2024-01-02</time><span class="post-meta-separator">|</span><i class="fas fa-history fa-fw post-meta-icon"></i><span class="post-meta-label">Updated</span><time class="post-meta-date-updated" datetime="2024-01-13T12:40:00.000Z" title="Updated 2024-01-13 20:40:00">2024-01-13</time></span><span class="post-meta-categories"><span class="post-meta-separator">|</span><i class="fas fa-inbox fa-fw post-meta-icon"></i><a class="post-meta-categories" href="/categories/RNA-seq-Report-Draft/">RNA-seq Report Draft</a></span></div><div class="meta-secondline"><span class="post-meta-separator">|</span><span class="post-meta-wordcount"><i class="far fa-file-word fa-fw post-meta-icon"></i><span class="post-meta-label">Word Count:</span><span class="word-count">3.8k</span><span class="post-meta-separator">|</span><i class="far fa-clock fa-fw post-meta-icon"></i><span class="post-meta-label">Reading Time:</span><span>12mins</span></span><span class="post-meta-separator">|</span><span class="post-meta-pv-cv" id="" data-flag-title="240102 Report Draft of DESeq2 WorkFlow"><i class="far fa-eye fa-fw post-meta-icon"></i><span class="post-meta-label">Post Views:</span><span id="busuanzi_value_page_pv"><i class="fa-solid fa-spinner fa-spin"></i></span></span></div></div></div><article class="post-content" id="article-container"><p>240102汇报口水稿，仅留档用，无演示，不建议阅读。</p>
<h2 id="相关内容"><a href="#相关内容" class="headerlink" title="相关内容"></a>相关内容</h2><p>PDF@6010_disk1_gmq: Report&#x2F;240102_DESeq2_WorkFlow.pdf<br>Code@6010_disk1_gmq: Code&#x2F;240102_DESeq2_WorkFlow.R<br>Link: <a target="_blank" rel="noopener" href="https://master.bioconductor.org/packages/release/workflows/vignettes/rnaseqGene/inst/doc/rnaseqGene.html">https://master.bioconductor.org/packages/release/workflows/vignettes/rnaseqGene/inst/doc/rnaseqGene.html</a></p>
<hr>
<h2 id="引入"><a href="#引入" class="headerlink" title="引入"></a>引入</h2><ul>
<li>今天是只关注这一个DESeq2的流程文档的介绍。</li>
<li>可能有一些细节的解释可能差点，因为我有一些没搞清楚就先跳。</li>
<li>这个文档是教我们如何从FASTQ文件出发来进行探索性数据分析和差异基因分析，并做出结果的可视化。所以它有一部分是在讲如何用Salmon处理得到量化到参考转录组的。我把这部分跳过去，直接从Salmon处理好的数据导入开始。</li>
<li>做探索性分析目的是观察我们的样本数据质量和样本之间的关系。</li>
<li>做差异分析则是为了找到在选定的实验条件下，不同组样本的基因表达具有显著性差异的基因。</li>
</ul>
<hr>
<p><strong>Contents, p1</strong></p>
<p>因此，我们的主要内容则包括这样几个步骤：</p>
<ul>
<li>首先简单介绍我们分析的数据，</li>
<li>如何导入数据，</li>
<li>然后是探索性分析，差异分析，可视化结果。</li>
<li>后面还会提及基因注释，移除隐藏的批次效应，时间过程实验。</li>
</ul>
<hr>
<p><strong>Experimental data, p2</strong></p>
<ul>
<li>我们的数据是从airway这个包里的数据，研究是的地塞米松对呼吸道平滑肌细胞的影响，有4个cell line，就是4个人作为细胞的供体。</li>
<li>因此我们就是会有8个样本，按照是否给地塞米松分为实验组和对照组，每一个组有4个不同供体的样本。</li>
</ul>
<hr>
<p><strong>Preparing quantification input to DESeq2, p2</strong></p>
<ul>
<li>我们输入的数据应当是一个没有做过normalize的counts矩阵，</li>
<li>前面经过salmon处理得到的是一个transcript-level quantification data，我们可以用tximport&#x2F;tximeta把他们aggregate到基因水平。</li>
<li>然后这里讲了一堆没太看懂的东西，我们先把这些跳过。</li>
<li>必须要知道的是，tximeta会得到一个SummarizedExperiment，这个是可以通过DESeqDataSet函数读到我们的DESeq2里面。</li>
</ul>
<hr>
<p><strong>Preparing libraries</strong></p>
<hr>
<h2 id="数据读取"><a href="#数据读取" class="headerlink" title="数据读取"></a>数据读取</h2><p><strong>Reading in data with tximeta, p4</strong></p>
<ul>
<li>首先找到要导入的counts数据，他们在quants文件夹下</li>
<li>我们还要导入一个对应包含有sample information的csv文件</li>
<li>把这两个数据都存到coldata里面，tximeta读取，summarizeToGene</li>
<li>就可以得到用Ensemble gene ID标识的基因表达数据了</li>
<li>如果原本的count矩阵存储形式不一样，也可以用其他的方式读取。</li>
</ul>
<hr>
<p><strong>SummarizedExperiment, p6</strong></p>
<ul>
<li><p>在这里简单介绍一下tximeta读取的SummarizedExperiment</p>
</li>
<li><p>包含3个部分：</p>
</li>
<li><p>assay：在我们这个例子里他包括了3个matrix：</p>
<ul>
<li>counts，是non-normalized counts，</li>
<li>abundance，是TPM数据。</li>
<li>length，代表基因的<strong>有效长度</strong>。（有关这个有效长度怎么理解我还说不好）<blockquote>
<p>TPM<br><em>Transcripts Per Kilobase of exon model per Million mapped reads</em></p>
</blockquote>
</li>
</ul>
</li>
<li><p>rowRanges：包含gene range信息，如基因的位置，正反读等。</p>
<ul>
<li>它还包括了一个metadata在seqinfo里面，是染色体的信息</li>
</ul>
</li>
<li><p>colData：包含sample的信息，如donor ID, treatment condition</p>
</li>
<li><p>因为我们需要的counts是assay的第一个matrix，直接用assay就行</p>
</li>
</ul>
<hr>
<p><strong>DESeqDataSet object, sample information, design formula, p9</strong></p>
<ul>
<li>DESeqDataSet build up on SummarizedExperiment, 有2个区别：<ol>
<li>DESeqDataSet的assay使用counts，其matrix value是非负整数</li>
<li>DESeqDataSet有一个design formula</li>
</ol>
</li>
<li>关于这个formula 的理解：还有点不是很透彻，目前我把他理解为：它是确定我们研究哪些因素来产生差异，比如group，treatment，我们可以把考虑到的因素都写进去，就写成~ group + treatment。后面还有提到的batch effect也可以写进去重新计算results。</li>
<li>它还可以有interaction terms，比如group: treatment，最后提到的的时间过程实验就是一种，它研究随时间的给药变化是否一致。</li>
</ul>
<hr>
<p><strong>Starting from SummarizedExperiment, p10</strong></p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">dds <span class="operator">&lt;-</span> DESeqDataSet<span class="punctuation">(</span>gse<span class="punctuation">,</span> design <span class="operator">=</span> <span class="operator">~</span> cell <span class="operator">+</span> dex<span class="punctuation">)</span></span><br></pre></td></tr></table></figure>
<p>注意确认样本信息的row是和count的column是对应正确的。</p>
<p><strong>Starting from count matrices, p10</strong></p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">ddsMat <span class="operator">&lt;-</span> DESeqDataSetFromMatrix<span class="punctuation">(</span></span><br><span class="line">	countData <span class="operator">=</span> countdata<span class="punctuation">,</span></span><br><span class="line">	colData <span class="operator">=</span> coldata<span class="punctuation">,</span></span><br><span class="line">	design <span class="operator">=</span> <span class="operator">~</span> cell <span class="operator">+</span> dex<span class="punctuation">)</span></span><br></pre></td></tr></table></figure>
<hr>
<h2 id="探索性分析与可视化"><a href="#探索性分析与可视化" class="headerlink" title="探索性分析与可视化"></a>探索性分析与可视化</h2><p>transformation of the counts → visually explotr sample relationship</p>
<ol>
<li>pre-filtering</li>
<li>VST and rlog</li>
</ol>
<ul>
<li>对于多维数据的探索性分析的一个常见方式是PCA，但是PCA通常是在方差变化很小的数据中表现的好。我们的RNA-seq数据的counts是<strong>variance会随着mean的增大而增大</strong>。<blockquote>
<p>Disperation<br>这里提到了RNA-seq的数据特征，所以再重新讲一负二项分布模型，以及纠正我之前的一个关于t检验的理解错误<br>【插入链接】<a target="_blank" rel="noopener" href="https://github.com/hbctraining/DGE_workshop_salmon/blob/master/lessons/01_DGE_setup_and_overview.md#:~:text=RNA%2Dseq%20count%20distribution">https://github.com/hbctraining/DGE_workshop_salmon/blob/master/lessons/01_DGE_setup_and_overview.md#:~:text=RNA%2Dseq%20count%20distribution</a><br>首先要介绍一下这个数据的特点：</p>
<ol>
<li>低counts数据占了很大部分</li>
<li>count的上限可以很高</li>
<li>数据的变化大。<br>（在这里它提到了微阵列数据的log是正态分布，但是，由于 RNA-seq 计数数据的不同特性，如整数计数而非连续测量值以及非正态分布数据，正态分布并不能准确地模拟 RNA-seq 计数。此处他讲的很清楚但我没太看懂，但我联想到了之前的问题：高通量测序不能用t检验。上次师兄也说我那个解释不太对，但我也没听懂师兄讲的和t检验有什么关系，所以我还是没懂，但这个不适合在这里展开，所以后面再说）</li>
</ol>
<p>回到这个数据这样的特点，我们讨论数据模型的选择，我们有二项分布、泊松分布、负二项分布等。</p>
<p>首先，p非常小，所以选择泊松分布或者负二项分布就比较合适。选择其中一种还是另一种将取决于我们数据的均值和方差之间的关系</p>
<p>第二，考虑到方差-均值关系，首先是方差是比均值要大的，其次是对应低表达的数据有很大的异方差性。因此我们选择了负二项分布。</p>
<p>最后，我们就不能选择泊松分布吗？不是的，如果我们的生物学重复足够多（比如＞20），那么方差就会减小逐渐逼近variance&#x3D;mean</p>
</blockquote>
</li>
<li>一个常见处理是log(counts+1)变换，但是这样会夸大低count的方差</li>
<li>我们这里介绍两种变换来解决这个问题：VST和rlog，如何选择：<ul>
<li>VST：速度快，异常值敏感性低，中到大数据集(n＞30)表现好</li>
<li>rlog：小数据集（n&lt;30）表现好，适合样本间测序深度变化大</li>
</ul>
</li>
<li>但是这个变换的时候声明了一个blind&#x3D;FALSE，我暂时<strong>没看懂</strong>这个参数的布尔值选择的原因。所以我选择跳过这里。</li>
<li>接下来就是对这三种方式的比较，log，VST，rlog。这里提到对于log的方式我们需要先对counts进行一个normalize来消除测序深度的影响。这就又回到了上次我没有解决的一个问题，DESeq2的normalize是怎样的，它的方法是比edgeR简单的，所以我很乐意在这里介绍一下，至于edgeR的比较复杂，我还没把它搬到今天的展示里面。不如下次再说。</li>
</ul>
<blockquote>
<p>Normalization<br>先求ln，然后对行求均值，把负无穷都移除，再把每一个ln值和行的ln均值做差，找到中位数。。。但是这个数学过程我有点记不清了，statquest就是讲了一下子</p>
</blockquote>
<ul>
<li>然后这里有个小问题会报错建议改一下就行，我还在这放了个总结</li>
</ul>
<hr>
<p><strong>Assess overall similarity between samples, p14</strong></p>
<ul>
<li>上面做的VST还有rlog都是为了处理也可以不只是给PCA用的，我们还有一个方式来估计样本间的相似性，就是直接计算欧氏距离</li>
<li>当然这个我具体就没有研究是怎么计算的了，我们直接用VST数据计算distance，然后可以画个热图来展示一下这个距离关系。</li>
<li>除此以外我们还可以计算泊松距离，原理也没看，注意他就不能用标准化之后的数据了，要用原始数据。也是可以画一个热图。</li>
<li>当然了，还有一个方法就是我们的常客PCA了，我们用VST数据就行了，嫌这个PCAplot丑的话我们还可以用ggplot，可以多打点标签。</li>
<li>GLM-PCA：输入count matrix，潜在维度&#x3D;2</li>
<li>MDS plot：多维标度图？只有距离的数据，没有matrix可以画这个</li>
</ul>
<hr>
<h2 id="差异分析"><a href="#差异分析" class="headerlink" title="差异分析"></a>差异分析</h2><ol>
<li>run the pipeline:<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">dds <span class="operator">&lt;-</span> DESeq<span class="punctuation">(</span>dds<span class="punctuation">)</span></span><br></pre></td></tr></table></figure></li>
<li>build the result table</li>
</ol>
<ul>
<li>默认会根据formula的最后一个变量来计算log2 fold change和p值</li>
<li>也可以通过加contrast参数来改变。这里是在解释这几个量是什么。</li>
<li>basemean：normalized count 的均值。</li>
<li>log2FoldChange：就是选定的formula变量之后计算的，lfcSE是标准误，我大概推测就是不同样本都算了一个lfc之后它就可以算这个样本的均值，然后和之前的lfc就能算标准误了。</li>
<li>这个stat就是统计量，Wald统计量这个问题，先看一眼这个是怎么算的：pdf，网址。</li>
<li>然后就是涉及到我的知识盲区了，我之前统计学的很粗糙而且都还给老师了，就是隐约记得z统计量他是在均值和方差都已知的时候算的，然后t是均值已知方差未知。然后都是正态分布。我们这里的正态分布应该是针对lfc来说的，不是count。那我对这个问题的认知就有点问题了，wald检验也是每次就检验这一个基因，所以也会存在需要FDR调整p值的问题。为什么不能用t检验呢。</li>
<li>这个wald test它就是在检验lfc是不是等于0，就这块反正还是挺迷茫的，首先lfc的分布就不是很理解，然后就要对他检验，所以实际上我们作图比如MA图里面的那个纵轴M和这个有什么样的联系呢？<blockquote>
<p>t-test<br>这个问题纠结了很久，大概找了一个看起来比较合适的解释的，但我没太理解，但再纠结下去我就在这个问题上耗太久了，就暂时放下了，在这我再念一下这个人的<a target="_blank" rel="noopener" href="https://www.biostars.org/p/376538/#:~:text=You%20can%20use,discussion%20of%20this).">解释</a>]。<br><em>“你可以使用 T 检验，limma 也是这么做的（不过是在将计数通过 voom 之后，再使用一些经验贝叶斯方法）。没有人对 RNA-seq 进行简单的 T 检验的原因与没有人对阵列数据进行 T 检验的原因相同，即在不汇集跨基因信息的情况下，很少有足够的样本数量来准确估计方差（有关这方面的讨论，请参阅 limma 的原始论文）。”</em><br>所以我感觉他的意思就是方差不准所以不用t检验，但我还不是特别理解为什么</p>
</blockquote>
</li>
<li>然后就是可以用summary来总结result</li>
<li>以及这里出现了p值是NA的情况，原因可能是有这样三点：count全是0，有异常值，自动的independent filter给筛掉了。这个independent filter在后面会提，这里就不说了</li>
<li>当然，也可以选formula里面其他的来比较</li>
<li>这里的padj就不再说了，他是用了那个BH方法，虽然我上次也没讲明白这个是怎么调整的，大概好像是他是给所有的p值排序，然后把这个p值乘上一个系数，总数除以他的排位，然后和排在这位的那个p比一下，选小的那个就是padj。</li>
</ul>
<hr>
<h2 id="结果展示"><a href="#结果展示" class="headerlink" title="结果展示"></a>结果展示</h2><p><strong>Counts plot</strong></p>
<p>对选定的p值最小的基因按组进行count的展示</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">topGene <span class="operator">&lt;-</span> rownames<span class="punctuation">(</span>res<span class="punctuation">)</span><span class="punctuation">[</span>which.min<span class="punctuation">(</span>res<span class="operator">$</span>padj<span class="punctuation">)</span><span class="punctuation">]</span></span><br><span class="line"><span class="comment"># 找到调整后的p值（padj）最小的基因在结果表（res）中的索引，从而找到最小padj对应的基因名</span></span><br><span class="line">plotCounts<span class="punctuation">(</span>dds<span class="punctuation">,</span> gene <span class="operator">=</span> topGene<span class="punctuation">,</span> intgroup<span class="operator">=</span><span class="built_in">c</span><span class="punctuation">(</span><span class="string">&quot;dex&quot;</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line"><span class="comment"># 在图中按照 &quot;dex&quot; 列指定的条件进行分组</span></span><br></pre></td></tr></table></figure>
<ul>
<li>除了直接展示counts以外，还可以用ggbeeswarm进行展示，此外，如果只展示普通点图的话，可以加上cell “line”</li>
<li>这几个图都很简单~直接看文档就行</li>
</ul>
<hr>
<p><strong>MAplot</strong></p>
<ul>
<li>做横轴为A：average of log；纵轴为M：minus of log的点图，把p值低的点用蓝色标记出来</li>
<li>作图注意的参数：coe &#x3D; resultNames(dds)[index]，type &#x3D; “apeglm”</li>
<li>作图常需要用shrink处理log，shrink的目的：更准确估计LFC。具体来说，是把所有基因log fold change估计值的分布作为先验，用于，将信息很少或高度分散的基因的log fold change估计值缩小，缩小到更有可能的较低的lfc估计值。</li>
<li>也可以在MA图中圈出你想强调的基因</li>
<li>另外一个图是p值的直方图，也可以很清楚的看到p值的频率分布来更好的了解数据</li>
</ul>
<hr>
<p><strong>Gene clustering</strong></p>
<ul>
<li>同样本聚类一样，基因也可以聚类，我们通过计算基因表达量与其平均值的差值来观察基因是高表达还是低表达</li>
<li>通过热图中的颜色就可以进行识别不同formula condition的基因表达是不同的</li>
<li>dendrogram则展示了基因间的树状分层关系<br>![[Pasted image 20240102225700.png]]</li>
</ul>
<hr>
<p><strong>Independent filtering</strong></p>
<ul>
<li>我们知道count量低的基因会更多的存在小p值的情况，但是这部分数据的噪音很大</li>
<li>因此，我们可以通过FDR 程序的输入中剔除低计数基因，发现更多具有显著性的基因，这样的过程会由result自动进行，也被称为independent filtering</li>
<li><strong>Independent Hypothesis Weighting</strong>（IHW）则是一种可以代替上述方法的方法，在执行result的时候修改filterFun &#x3D; ihw即可</li>
</ul>
<hr>
<h2 id="结果注释与导出"><a href="#结果注释与导出" class="headerlink" title="结果注释与导出"></a>结果注释与导出</h2><ul>
<li>注释：我们可以根据Ensembl gene IDs来找到基因名，从而对基因用<code>mapIDs()</code>进行注释</li>
<li>导出：我们可以导出csv文件<code>write.csv</code>，动态html文件<code>HTMLReport()</code></li>
<li>在基因组空间展示fold changes：通过<code>format</code>来获取gene range的情况，通过<code>Gviz</code>作图。主要过程：<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br></pre></td><td class="code"><pre><span class="line">window <span class="operator">&lt;-</span> resGR<span class="punctuation">[</span>topGene<span class="punctuation">]</span> <span class="operator">+</span> <span class="number">1e6</span></span><br><span class="line"><span class="comment"># 先创建一个Window of gene with the smallest p value, 包括它上下的1 million base pairs</span></span><br><span class="line">strand<span class="punctuation">(</span>window<span class="punctuation">)</span> <span class="operator">&lt;-</span> <span class="string">&quot;*&quot;</span></span><br><span class="line">resGRsub <span class="operator">&lt;-</span> resGR<span class="punctuation">[</span>resGR <span class="operator">%over%</span> window<span class="punctuation">]</span></span><br><span class="line">naOrDup <span class="operator">&lt;-</span> <span class="built_in">is.na</span><span class="punctuation">(</span>resGRsub<span class="operator">$</span>symbol<span class="punctuation">)</span> <span class="operator">|</span> duplicated<span class="punctuation">(</span>resGRsub<span class="operator">$</span>symbol<span class="punctuation">)</span></span><br><span class="line">resGRsub<span class="operator">$</span>group <span class="operator">&lt;-</span> ifelse<span class="punctuation">(</span>naOrDup<span class="punctuation">,</span> <span class="built_in">names</span><span class="punctuation">(</span>resGRsub<span class="punctuation">)</span><span class="punctuation">,</span> resGRsub<span class="operator">$</span>symbol<span class="punctuation">)</span></span><br><span class="line"><span class="comment"># 判断这个symbol在subset中存在，且不是duplicated我们就把这个gene symbol添加为resGRsub的name</span></span><br><span class="line">status <span class="operator">&lt;-</span> factor<span class="punctuation">(</span>ifelse<span class="punctuation">(</span>resGRsub<span class="operator">$</span>padj <span class="operator">&lt;</span> <span class="number">0.05</span> <span class="operator">&amp;</span> <span class="operator">!</span><span class="built_in">is.na</span><span class="punctuation">(</span>resGRsub<span class="operator">$</span>padj<span class="punctuation">)</span><span class="punctuation">,</span><span class="string">&quot;sig&quot;</span><span class="punctuation">,</span> <span class="string">&quot;notsig&quot;</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line"><span class="comment"># 然后找出来这个subset里面padj小的，就说它的status factor = &quot;sig&quot;, else status factor = &quot;notsig&quot;</span></span><br><span class="line"><span class="comment"># 创建3个track，绘图即可</span></span><br><span class="line">options<span class="punctuation">(</span>ucscChromosomeNames <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">)</span></span><br><span class="line">g <span class="operator">&lt;-</span> GenomeAxisTrack<span class="punctuation">(</span><span class="punctuation">)</span></span><br><span class="line">a <span class="operator">&lt;-</span> AnnotationTrack<span class="punctuation">(</span>resGRsub<span class="punctuation">,</span> name <span class="operator">=</span> <span class="string">&quot;gene ranges&quot;</span><span class="punctuation">,</span> feature <span class="operator">=</span> status<span class="punctuation">)</span></span><br><span class="line">d <span class="operator">&lt;-</span> DataTrack<span class="punctuation">(</span>resGRsub<span class="punctuation">,</span> data <span class="operator">=</span> <span class="string">&quot;log2FoldChange&quot;</span><span class="punctuation">,</span> baseline <span class="operator">=</span> <span class="number">0</span><span class="punctuation">,</span></span><br><span class="line">type <span class="operator">=</span> <span class="string">&quot;h&quot;</span><span class="punctuation">,</span> name <span class="operator">=</span> <span class="string">&quot;log2 fold change&quot;</span><span class="punctuation">,</span> strand <span class="operator">=</span> <span class="string">&quot;+&quot;</span><span class="punctuation">)</span></span><br><span class="line">plotTracks<span class="punctuation">(</span><span class="built_in">list</span><span class="punctuation">(</span>g<span class="punctuation">,</span> d<span class="punctuation">,</span> a<span class="punctuation">)</span><span class="punctuation">,</span> groupAnnotation <span class="operator">=</span> <span class="string">&quot;group&quot;</span><span class="punctuation">,</span></span><br><span class="line">notsig <span class="operator">=</span> <span class="string">&quot;grey&quot;</span><span class="punctuation">,</span> sig <span class="operator">=</span> <span class="string">&quot;hotpink&quot;</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure></li>
</ul>
<hr>
<p><strong>Removing hidden batch effects</strong></p>
<ul>
<li>不同cell lines可能会产生batch effects，我们可以用 SVA&#x2F;RUVSeq 的 statistical methods检测，然后把这个加到 DESeqDataSet 的design里面来解释他们</li>
<li>他们用的term：<ul>
<li>SVA: surrogate variables</li>
<li>RUV: unwanted variation</li>
</ul>
</li>
</ul>
<hr>
<p><strong>Time course experiments</strong></p>
</article><div class="post-copyright"><div class="post-copyright__author"><span class="post-copyright-meta"><i class="fas fa-circle-user fa-fw"></i>Author: </span><span class="post-copyright-info"><a href="http://www.monig.cn">Mengqi</a></span></div><div class="post-copyright__type"><span class="post-copyright-meta"><i class="fas fa-square-arrow-up-right fa-fw"></i>Link: </span><span class="post-copyright-info"><a href="http://www.monig.cn/2024/01/02/240102%20Report%20Draft%20of%20DESeq2%20WorkFlow/">http://www.monig.cn/2024/01/02/240102%20Report%20Draft%20of%20DESeq2%20WorkFlow/</a></span></div><div class="post-copyright__notice"><span class="post-copyright-meta"><i class="fas fa-circle-exclamation fa-fw"></i>Copyright Notice: </span><span class="post-copyright-info">All articles in this blog are licensed under <a target="_blank" rel="noopener" href="https://creativecommons.org/licenses/by-nc-sa/4.0/">CC BY-NC-SA 4.0</a> unless stating additionally.</span></div></div><div class="tag_share"><div class="post-meta__tag-list"><a class="post-meta__tags" href="/tags/Not-Recommended-Reading/">Not Recommended Reading</a></div><div class="post_share"><div class="social-share" data-image="https://s2.loli.net/2024/01/14/BlD6LX3CoIqpn4x.jpg" data-sites="facebook,twitter,wechat,weibo,qq"></div><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/butterfly-extsrc@1.1.3/sharejs/dist/css/share.min.css" media="print" onload="this.media='all'"><script src="https://cdn.jsdelivr.net/npm/butterfly-extsrc@1.1.3/sharejs/dist/js/social-share.min.js" defer></script></div></div><nav class="pagination-post" id="pagination"><div class="prev-post pull-full"><a href="/2024/01/12/240112%20Report%20Draft%20of%20Upstream/" title="240112 Report Draft of Upstream"><img class="cover" src="https://s2.loli.net/2024/01/14/BlD6LX3CoIqpn4x.jpg" onerror="onerror=null;src='/img/404.jpg'" alt="cover of previous post"><div class="pagination-info"><div class="label">Previous</div><div class="prev_info">240112 Report Draft of Upstream</div></div></a></div></nav><div class="relatedPosts"><div class="headline"><i class="fas fa-thumbs-up fa-fw"></i><span>Related Articles</span></div><div class="relatedPosts-list"><div><a href="/2024/01/12/240112%20Report%20Draft%20of%20Upstream/" title="240112 Report Draft of Upstream"><img class="cover" src="https://s2.loli.net/2024/01/14/BlD6LX3CoIqpn4x.jpg" alt="cover"><div class="content is-center"><div class="date"><i class="far fa-calendar-alt fa-fw"></i> 2024-01-12</div><div class="title">240112 Report Draft of Upstream</div></div></a></div></div></div></div><div class="aside-content" id="aside-content"><div class="card-widget card-info"><div class="is-center"><div class="avatar-img"><img src="https://s2.loli.net/2024/01/14/Y51RcC3rJjgFDkS.jpg" onerror="this.onerror=null;this.src='/img/friend_404.gif'" alt="avatar"/></div><div class="author-info__name">Mengqi</div><div class="author-info__description">Keep going, the journey is the reward!<p></div></div><div class="card-info-data site-data is-center"><a href="/archives/"><div class="headline">Articles</div><div class="length-num">2</div></a><a href="/tags/"><div class="headline">Tags</div><div class="length-num">1</div></a><a href="/categories/"><div class="headline">Categories</div><div class="length-num">1</div></a></div><div class="card-info-social-icons is-center"><a class="social-icon" href="mailto:gaomengqi_01@163.com" target="_blank" title="Email"><i class="fas fa-envelope" style="color: #4a7dbe;"></i></a></div></div><div class="sticky_layout"><div class="card-widget" id="card-toc"><div class="item-headline"><i class="fas fa-stream"></i><span>Contents</span><span class="toc-percentage"></span></div><div class="toc-content"><ol class="toc"><li class="toc-item toc-level-2"><a class="toc-link" href="#%E7%9B%B8%E5%85%B3%E5%86%85%E5%AE%B9"><span class="toc-number">1.</span> <span class="toc-text">相关内容</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E5%BC%95%E5%85%A5"><span class="toc-number">2.</span> <span class="toc-text">引入</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E6%95%B0%E6%8D%AE%E8%AF%BB%E5%8F%96"><span class="toc-number">3.</span> <span class="toc-text">数据读取</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E6%8E%A2%E7%B4%A2%E6%80%A7%E5%88%86%E6%9E%90%E4%B8%8E%E5%8F%AF%E8%A7%86%E5%8C%96"><span class="toc-number">4.</span> <span class="toc-text">探索性分析与可视化</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E5%B7%AE%E5%BC%82%E5%88%86%E6%9E%90"><span class="toc-number">5.</span> <span class="toc-text">差异分析</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E7%BB%93%E6%9E%9C%E5%B1%95%E7%A4%BA"><span class="toc-number">6.</span> <span class="toc-text">结果展示</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E7%BB%93%E6%9E%9C%E6%B3%A8%E9%87%8A%E4%B8%8E%E5%AF%BC%E5%87%BA"><span class="toc-number">7.</span> <span class="toc-text">结果注释与导出</span></a></li></ol></div></div><div class="card-widget card-recent-post"><div class="item-headline"><i class="fas fa-history"></i><span>Recent Post</span></div><div class="aside-list"><div class="aside-list-item"><a class="thumbnail" href="/2024/01/12/240112%20Report%20Draft%20of%20Upstream/" title="240112 Report Draft of Upstream"><img src="https://s2.loli.net/2024/01/14/BlD6LX3CoIqpn4x.jpg" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="240112 Report Draft of Upstream"/></a><div class="content"><a class="title" href="/2024/01/12/240112%20Report%20Draft%20of%20Upstream/" title="240112 Report Draft of Upstream">240112 Report Draft of Upstream</a><time datetime="2024-01-12T12:30:00.000Z" title="Created 2024-01-12 20:30:00">2024-01-12</time></div></div><div class="aside-list-item"><a class="thumbnail" href="/2024/01/02/240102%20Report%20Draft%20of%20DESeq2%20WorkFlow/" title="240102 Report Draft of DESeq2 WorkFlow"><img src="https://s2.loli.net/2024/01/14/BlD6LX3CoIqpn4x.jpg" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="240102 Report Draft of DESeq2 WorkFlow"/></a><div class="content"><a class="title" href="/2024/01/02/240102%20Report%20Draft%20of%20DESeq2%20WorkFlow/" title="240102 Report Draft of DESeq2 WorkFlow">240102 Report Draft of DESeq2 WorkFlow</a><time datetime="2024-01-02T12:30:00.000Z" title="Created 2024-01-02 20:30:00">2024-01-02</time></div></div></div></div></div></div></main><footer id="footer"><div id="footer-wrap"><div class="copyright">&copy;2024 By Mengqi</div><div class="footer_custom_text">“Learning is the only thing the mind never exhausts, never fears, and never regrets.” -- Leonardo da Vinci.<p><a target="_blank" href="https://hexo.io/"><img src="https://img.shields.io/badge/Frame-Hexo-blue?style=flat&logo=hexo" title="博客框架为Hexo"></a>&nbsp;<a target="_blank" href="https://butterfly.js.org/"><img src="https://img.shields.io/badge/Theme-Butterfly-6513df?style=flat&logo=bitdefender" title="主题采用butterfly"></a>&nbsp;<a target="_blank" href="https://vercel.com/ "><img src="https://img.shields.io/badge/Hosted-Vervel-brightgreen?style=flat&logo=Vercel" title="本站采用双线部署，默认线路托管于Vercel"></a>&nbsp;<a target="_blank" href="https://github.com/"><img src="https://img.shields.io/badge/Source-Github-d021d6?style=flat&logo=GitHub" title="本站项目由Gtihub托管"></a>&nbsp;<a target="_blank" href="http://creativecommons.org/licenses/by-nc-sa/4.0/"><img src="https://img.shields.io/badge/Copyright-BY--NC--SA%204.0-d42328?style=flat&logo=Claris" title="本站采用知识共享署名-非商业性使用-相同方式共享4.0国际许可协议进行许可"></a></p></div></div></footer></div><div id="rightside"><div id="rightside-config-hide"><button id="readmode" type="button" title="Read Mode"><i class="fas fa-book-open"></i></button><button id="darkmode" type="button" title="Toggle Between Light And Dark Mode"><i class="fas fa-adjust"></i></button><button id="hide-aside-btn" type="button" title="Toggle between Single-column and Double-column"><i class="fas fa-arrows-alt-h"></i></button></div><div id="rightside-config-show"><button id="rightside-config" type="button" title="Setting"><i class="fas fa-cog fa-spin"></i></button><button class="close" id="mobile-toc-button" type="button" title="Table Of Contents"><i class="fas fa-list-ul"></i></button><button id="go-up" type="button" title="Back To Top"><span class="scroll-percent"></span><i class="fas fa-arrow-up"></i></button></div></div><div><script src="/js/utils.js?v=4.12.0"></script><script src="/js/main.js?v=4.12.0"></script><script src="https://cdn.jsdelivr.net/npm/@fancyapps/ui@5.0.32/dist/fancybox/fancybox.umd.min.js"></script><div class="js-pjax"></div><script src="https://cdn.jsdelivr.net/npm/pjax@0.2.8/pjax.min.js"></script><script>let pjaxSelectors = ["head > title","#config-diff","#body-wrap","#rightside-config-hide","#rightside-config-show",".js-pjax"]

var pjax = new Pjax({
  elements: 'a:not([target="_blank"])',
  selectors: pjaxSelectors,
  cacheBust: false,
  analytics: false,
  scrollRestoration: false
})

document.addEventListener('pjax:send', function () {

  // removeEventListener
  btf.removeGlobalFnEvent('pjax')
  btf.removeGlobalFnEvent('themeChange')

  document.getElementById('rightside').classList.remove('rightside-show')
  
  if (window.aplayers) {
    for (let i = 0; i < window.aplayers.length; i++) {
      if (!window.aplayers[i].options.fixed) {
        window.aplayers[i].destroy()
      }
    }
  }

  typeof typed === 'object' && typed.destroy()

  //reset readmode
  const $bodyClassList = document.body.classList
  $bodyClassList.contains('read-mode') && $bodyClassList.remove('read-mode')

  typeof disqusjs === 'object' && disqusjs.destroy()
})

document.addEventListener('pjax:complete', function () {
  window.refreshFn()

  document.querySelectorAll('script[data-pjax]').forEach(item => {
    const newScript = document.createElement('script')
    const content = item.text || item.textContent || item.innerHTML || ""
    Array.from(item.attributes).forEach(attr => newScript.setAttribute(attr.name, attr.value))
    newScript.appendChild(document.createTextNode(content))
    item.parentNode.replaceChild(newScript, item)
  })

  GLOBAL_CONFIG.islazyload && window.lazyLoadInstance.update()

  typeof panguInit === 'function' && panguInit()

  // google analytics
  typeof gtag === 'function' && gtag('config', '', {'page_path': window.location.pathname});

  // baidu analytics
  typeof _hmt === 'object' && _hmt.push(['_trackPageview',window.location.pathname]);

  typeof loadMeting === 'function' && document.getElementsByClassName('aplayer').length && loadMeting()

  // prismjs
  typeof Prism === 'object' && Prism.highlightAll()
})

document.addEventListener('pjax:error', e => {
  if (e.request.status === 404) {
    pjax.loadUrl('/404.html')
  }
})</script><script async data-pjax src="//busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script></div></body></html>